## References

*Xi Zhang, David Roy Smith,John M Archibald,  Key step missing for input file preparation in MCScanX , Nature protocols, 2025;, (in submission)*

### References for Dependencies 
1. **McScanX_protocol**: Wang, Yupeng, et al. "Detection of colinear blocks and synteny and evolutionary analyses based on utilization of MCScanX." Nature Protocols 19.7 (2024): 2206-2229.
2. **McScanX**: Wang, Yupeng, et al. "MCScanX: a toolkit for detection and evolutionary analysis of gene synteny and collinearity." Nucleic acids research 40.7 (2012): e49-e49.
3. **DupGen_finder**: Qiao, Xin, et al. "Gene duplication and evolution in recurring polyploidizationâ€“diploidization cycles in plants." Genome biology 20 (2019): 1-23.
4. **BEDOPS** Neph, Shane, et al. "BEDOPS: high-performance genomic feature operations." Bioinformatics 28.14 (2012): 1919-1920
### Links to the Dependencies:
 
 1. Diamond: https://github.com/bbuchfink/diamond
 2. McScanX_protocol: http://bdx-consulting.com/mcscanx-protocol/
 3. DupGen_finder: https://github.com/qiao-xin/DupGen_finder/tree/master
 4. McScanX: https://github.com/wyp1125/MCScanX
 5. BEDOPS v2.4.4: https://bedops.readthedocs.io/en/latest/
