## References

*Xi Zhang, Yining Hu, David Roy Smith, Zhenyu Cheng, John M Archibald, HSDSnake: a user-friendly SnakeMake pipeline for analysis of duplicate genes in eukaryotic genomes, Bioinformatics, 2025;, btaf325, https://doi.org/10.1093/bioinformatics/btaf325*

### References for Dependencies 
1. **McScanX_protocol**: Wang, Yupeng, et al. "Detection of colinear blocks and synteny and evolutionary analyses based on utilization of MCScanX." Nature Protocols 19.7 (2024): 2206-2229.
2. **McScanX**: Wang, Yupeng, et al. "MCScanX: a toolkit for detection and evolutionary analysis of gene synteny and collinearity." Nucleic acids research 40.7 (2012): e49-e49.
3. **DupGen_finder**: Qiao, Xin, et al. "Gene duplication and evolution in recurring polyploidizationâ€“diploidization cycles in plants." Genome biology 20 (2019): 1-23.
4. **KEGG_BlastKOALA**: Kanehisa, Minoru, Yoko Sato, and Kanae Morishima. "BlastKOALA and GhostKOALA: KEGG tools for functional characterization of genome and metagenome sequences." Journal of molecular biology 428.4 (2016): 726-731.
5. **InterProScan**: Jones, Philip, et al. "InterProScan 5: genome-scale protein function classification." Bioinformatics 30.9 (2014): 1236-1240.
6. **PAML**: Yang, Ziheng. "PAML 4: phylogenetic analysis by maximum likelihood." Molecular biology and evolution 24.8 (2007): 1586-1591.
7. **HSDFinder_protocol**: Xi Zhang, Yining Hu, David Roy Smith. (2021). HSDFinder: a BLAST-based strategy to search for highly similar duplicated genes in eukaryotic genomes. Frontiers in Bioinformatics. doi: http://doi.org/10.3389/fbinf.2021.803176
8. **HSDFinder**: Xi Zhang, Yining Hu, David Roy Smith. (2021). HSDFinder: a BLAST-based strategy to search for highly similar duplicated genes in eukaryotic genomes. Frontiers in Bioinformatics. doi: http://doi.org/10.3389/fbinf.2021.803176
9. **HSDecipher**: Xi Zhang, Yining Hu, Zhenyu Cheng, John M. Archibald (2023). HSDecipher: A pipeline for comparative genomic analysis of highly similar duplicate genes in eukaryotic genomes. StarProtocols. doi:  doi: https://doi.org/10.1016/j.xpro.2022.102014

### Other relevant References

1. Zhang, X., Hu, Y. & Smith, D. R. 2022. HSDatabase - a database of highly similar duplicate genes from plants, animals, and algae. Database, doi:http://doi.org/10.1093/database/baac086.
2. Zhang, X. & Smith, D. R. 2022. An overview of online resources for intra-species detection of gene duplications. Frontiers in Genetics, doi: http://doi.org/10.3389/fgene.2022.1012788.
3. Xi Zhang, et.al. David Roy Smith (2021). Draft genome sequence of the Antarctic green alga Chlamydomonas sp. UWO241 DOI:https://doi.org/10.1016/j.isci.2021.102084

### Links to the Dependencies:
 
 1. Pfam 37.0 (Sep 2024, 21,979 entries): https://pfam.xfam.org
 2. InterPro 101.0 (Jul 2024, 45,899 entries):http://www.ebi.ac.uk/interpro/
 3. KEGG Orthology Database: https://www.genome.jp/kegg/ko.html
 4. InterProscan: https://github.com/ebi-pf-team/interproscan
 5. KEGG : https://www.kegg.jp/kegg/
 6. Diamond: https://github.com/bbuchfink/diamond
 7. McScanX_protocol: http://bdx-consulting.com/mcscanx-protocol/
 8. DupGen_finder: https://github.com/qiao-xin/DupGen_finder/tree/master
 9. McScanX: https://github.com/wyp1125/MCScanX
